File: api/scalars/organization_acronym.py
from re import compile
from graphene.types import Scalar
from graphql.language import ast
from graphql import GraphQLError

from functions.error_messages import *

ACRONYM_REGEX = r'''^[A-Z0-9_]{1,10}$'''
ACRONYM_REGEX = compile(ACRONYM_REGEX)


class Acronym(Scalar):
    """
    A field whose value is an upper case letter or an under score that has a
    length between 1 and 10
    """
    @staticmethod
    def serialize(value):
        if not isinstance(value, str):
            raise GraphQLError(scalar_error_type("String", value))

        if not ACRONYM_REGEX.search(value):
            raise GraphQLError(scalar_error_type("Acronym", value))
        return value

    @staticmethod
    def parse_value(value):
        if not isinstance(value, str):
            raise GraphQLError(scalar_error_type('String', value))

        if not ACRONYM_REGEX.search(value):
            raise GraphQLError(scalar_error_type("String", value))
        return value

    @staticmethod
    def parse_literal(node):
        if not isinstance(node, ast.StringValue):
            raise GraphQLError(scalar_error_only_types("strings", "acronym", str(ast.Type)))

        if not ACRONYM_REGEX.search(node.value):
            raise GraphQLError(scalar_error_type("Acronym", node.value))
        return node.value

File: api/src/scalars/organization-acronym.js
import {Kind, GraphQLError, GraphQLScalarType} from 'graphql'

const validate = (value) => {
  const ACRONYM_REGEX = /^[A-Za-z0-9_-]{1,50}$/

  if (typeof value !== 'string') {
    throw new TypeError(`Value is not string: ${typeof value}`)
  }

  if (!ACRONYM_REGEX.test(value)) {
    throw new TypeError(`Value is not a valid acronym: ${value}`)
  }
  return value
}

export const Acronym = new GraphQLScalarType({
  name: 'Acronym',
  description:
    'A field whose value consists of upper case or lower case letters or underscores with a length between 1 and 50.',
  serialize: validate,
  parseValue: validate,

  parseLiteral(ast) {
    if (ast.kind !== Kind.STRING) {
      throw new GraphQLError(
        `Can only validate strings as acronyms but got a: ${ast.kind}`,
      )
    }
    return validate(ast.value)
  },
})
